### PR TITLE
QAston Proc System

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:
-  
-  

## Issues Addressed:

- Closes https://github.com/chromiecraft/chromiecraft/issues/3269
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9177

## Tests Performed:

- Not Tested
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by temperrr** [comment]
Please tell us how to test it before marking it "waiting to be tested" Thanks!

**Comment by Lynethris** [comment]
Error during the build:

Screenshot 2022-03-31 202543

**Comment by IntelligentQuantum** [comment]
@Lynethris Test Please

**Comment by temperrr** [comment]
Breaks eluna

```
In file included from /mnt/data/azerothcore/modules/mod-eluna/src/LuaEngine/LuaFunctions.cpp:23:
/mnt/data/azerothcore/modules/mod-eluna/src/LuaEngine/UnitMethods.h:2827:24: fatal error: no matching constructor for initialization of 'DamageInfo'
            DamageInfo dmgInfo(unit, target, damage, nullptr, schoolmask, SPELL_DIRECT_DAMAGE);
                       ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:766:5: note: candidate constructor not viable: requires 4 arguments, but 6 were provided
    DamageInfo(SpellNonMeleeDamage const& spellNonMeleeDamage, DamageEffectType damageType, WeaponAttackType attackType, uint32 hitMask);
    ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:764:5: note: candidate constructor not viable: requires at least 7 arguments, but 6 were provided
    DamageInfo(Unit* attacker, Unit* victim, uint32 damage, SpellInfo const* spellInfo, SpellSchoolMask schoolMask, DamageEffectType damageType, WeaponAttackType attackType, uint32 cleanDamage = 0);
    ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:765:14: note: candidate constructor not viable: requires single argument 'dmgInfo', but 6 arguments were provided
    explicit DamageInfo(CalcDamageInfo& dmgInfo);
             ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:748:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 6 were provided
class DamageInfo
      ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:748:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 6 were provided
2 warnings and 1 error generated.
```

Testing further without eluna bit this need fix

**Comment by temperrr** [comment]
Breaks Eluna
```
In file included from /mnt/data/azerothcore/modules/mod-eluna/src/LuaEngine/LuaFunctions.cpp:23:
/mnt/data/azerothcore/modules/mod-eluna/src/LuaEngine/UnitMethods.h:2827:24: fatal error: no matching constructor for initialization of 'DamageInfo'
            DamageInfo dmgInfo(unit, target, damage, nullptr, schoolmask, SPELL_DIRECT_DAMAGE);
                       ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:766:5: note: candidate constructor not viable: requires 4 arguments, but 6 were provided
    DamageInfo(SpellNonMeleeDamage const& spellNonMeleeDamage, DamageEffectType damageType, WeaponAttackType attackType, uint32 hitMask);
    ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:764:5: note: candidate constructor not viable: requires at least 7 arguments, but 6 were provided
    DamageInfo(Unit* attacker, Unit* victim, uint32 damage, SpellInfo const* spellInfo, SpellSchoolMask schoolMask, DamageEffectType damageType, WeaponAttackType attackType, uint32 cleanDamage = 0);
    ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:765:14: note: candidate constructor not viable: requires single argument 'dmgInfo', but 6 arguments were provided
    explicit DamageInfo(CalcDamageInfo& dmgInfo);
             ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:748:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 6 were provided
class DamageInfo
      ^
/mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:748:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 6 were provided
2 warnings and 1 error generated.```

**Comment by temperrr** [comment]
Stormstrike (shaman talent) nature increased damage taken does not get used by shaman nature ability (example: earth shock, lightning bolt)

**Comment by IntelligentQuantum** [comment]
> Breaks Eluna
> 
> ```
> In file included from /mnt/data/azerothcore/modules/mod-eluna/src/LuaEngine/LuaFunctions.cpp:23:
> /mnt/data/azerothcore/modules/mod-eluna/src/LuaEngine/UnitMethods.h:2827:24: fatal error: no matching constructor for initialization of 'DamageInfo'
>             DamageInfo dmgInfo(unit, target, damage, nullptr, schoolmask, SPELL_DIRECT_DAMAGE);
>                        ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:766:5: note: candidate constructor not viable: requires 4 arguments, but 6 were provided
>     DamageInfo(SpellNonMeleeDamage const& spellNonMeleeDamage, DamageEffectType damageType, WeaponAttackType attackType, uint32 hitMask);
>     ^
> /mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:764:5: note: candidate constructor not viable: requires at least 7 arguments, but 6 were provided
>     DamageInfo(Unit* attacker, Unit* victim, uint32 damage, SpellInfo const* spellInfo, SpellSchoolMask schoolMask, DamageEffectType damageType, WeaponAttackType attackType, uint32 cleanDamage = 0);
>     ^
> /mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:765:14: note: candidate constructor not viable: requires single argument 'dmgInfo', but 6 arguments were provided
>     explicit DamageInfo(CalcDamageInfo& dmgInfo);
>              ^
> /mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:748:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 6 were provided
> class DamageInfo
>       ^
> /mnt/data/azerothcore/src/server/game/Entities/Unit/Unit.h:748:7: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 6 were provided
> 2 warnings and 1 error generated. 
> ```
> 
> and warnings:
> 
> image

I know warnings and eluna are not important right now

**Comment by temperrr** [comment]
warrior: Talent sword and broad does not refresh shield slam cooldown when on cooldown:

**Comment by IntelligentQuantum** [comment]
@temperrr Is there another problem?

**Comment by temperrr** [comment]
We are still testing other classes

**Comment by Lynethris** [comment]
I will start this afternoon to test priest, mage and warlock. Sorry, I wanted to finish before "Implementation save instance gameobject".

**Comment by temperrr** [comment]
Druid:

Balance:
Improved Faerie fire extra crit and hit chance does not seem to work? (no entry in list aura)
nature's grace does not trigger
Eclipse does not trigger

Feral:
Improved Leader of Pack does not give mana or hp.

Warlock:

Demonic Pact is applying on every hit instead of every crit from pet.
Molten Core does not proc of corruption

**Comment by Lynethris** [comment]
-Priest: 
Shadow talent: Shadow weaving - Doesn’t proc

Mage: All seems work.

**Comment by Lynethris** [comment]
DK Seems good too

**Comment by temperrr** [comment]
Whats the status? :)

**Comment by IntelligentQuantum** [comment]
I will work on this on Friday

**Comment by acidmanifesto** [comment]
What is the current status of this? Is it abandoned?

**Comment by IntelligentQuantum** [comment]
> What is the current status of this? Is it abandoned?

Nope

**Comment by DavuKnight** [comment]
Currently tagged as "Waiting to be Tested"
But there are no testing steps and there are conflicts
Should this be moved to WIP instead?

**Comment by Maelthyrr** [comment]
You can look above for tests. But globaly, need to check every spells and talents. It's a global system. You don't have standard progress to test it.

Merge conflicts should be for the merge on master branch. With the standard process, it should be fine. If You have an issue reply here for an update.

**Comment by Maelthyrr** [comment]
@IntelligentQuantum I had to recompile but now I got merge conflicts. Could you update this to finish my test ?

**Comment by IntelligentQuantum** [comment]
@Maelthyrr Done

**Comment by Maelthyrr** [comment]
Strange I got many issues with the sql querie: 
- Duplicate entry '71201-spell_igb_battle_experience_check' for key 'spell_script_names.spell_id'
- Some columns unknow in spell_procs: typemask, ratePerMinute... and

**Comment by IntelligentQuantum** [comment]
> Strange I got many issues with the sql querie (I've drop my db before):
> 
>     * Duplicate entry '71201-spell_igb_battle_experience_check' for key 'spell_script_names.spell_id'
> 
>     * Some columns unknow in spell_procs: typemask, ratePerMinute...

@Maelthyrr Done

**Comment by Maelthyrr** [comment]
@IntelligentQuantum tested redone, and I've compare to CC PTR to be sure that it's related to this PR:

Some worlserver errors: 
- [ ] AuraEffect::HandleProcTriggerSpellAuraProc: Could not trigger spell 0 from aura 24932 proc, because the spell does not have an entry in Spell.dbc.
- [ ] AuraEffect::HandleProcTriggerSpellAuraProc: Could not trigger spell 0 from aura 48525 proc, because the spell does not have an entry in Spell.dbc.

Warlock issues:
- [ ] Molten core (talent) can proc without corruption
- [ ] Demonic Pact doesn't proc on a crit pet

Druid:
- [ ] Nature Grace doesn't proc (balance talent)
- [ ] Eclispe doesn't proc (balance talent)
- [ ] Improve faerie doesn't work (balance talent)
- [ ] Improve Leader of Pack does not give mana or hp (feral talent)

Shaman:
- [ ] Stormstrike doesn't apply damage bonus from elemental spells

**Comment by IntelligentQuantum** [comment]
> @IntelligentQuantum tested redone, and I've compare to CC PTR to be sure that it's related to this PR: Almost same issues than before.
> 
> Some worlserver errors:
> 
>     * [ ]  AuraEffect::HandleProcTriggerSpellAuraProc: Could not trigger spell 0 from aura 24932 proc, because the spell does not have an entry in Spell.dbc.
> 
>     * [ ]  AuraEffect::HandleProcTriggerSpellAuraProc: Could not trigger spell 0 from aura 48525 proc, because the spell does not have an entry in Spell.dbc.
> 
> 
> Warlock issues:
> 
>     * [ ]  Molten core (talent) can proc without corruption
> 
>     * [ ]  Demonic Pact doesn't proc on a crit pet
> 
> 
> Druid:
> 
>     * [ ]  Nature Grace doesn't proc (balance talent)
> 
>     * [ ]  Eclispe doesn't proc (balance talent)
> 
>     * [ ]  Improve faerie doesn't work (balance talent)
> 
>     * [ ]  Improve Leader of Pack does not give mana or hp (feral talent)
> 
> 
> Shaman:
> 
>     * [ ]  Stormstrike doesn't apply damage bonus from elemental spells
> 
> 
> Paladin:
> 
>     * [ ]  Sacred shield aura will kill the player if he takes damage
> 
> 
> Otherwise any new issues find.

These bugs will be fixed in the new PRS.

Now this PR needs a code review, and if there is a specific problem, it should be fixed and merged quickly

Please spend time on this if you have time @Nefertumm

**Comment by Kitzunu** [comment]
yea impossible to fix all the issues in one PR. 

Pro's: We get a new updated good system
Con's: We will get some broken spells.

But that is well worth it.

File: src/server/scripts/Outland/boss_doomlord_kazzak.cpp [hunk 1]
```diff
@@ -219,8 +220,47 @@ class spell_mark_of_kazzak : public SpellScriptLoader
...
+    {
+        return new spell_twisted_reflection_AuraScript();
+    }
+};
+
 void AddSC_boss_doomlordkazzak()
 {
     new boss_doomlord_kazzak();
     new spell_mark_of_kazzak();
+    new spell_twisted_reflection();
```
Comment (Kitzunu) [inline src/server/scripts/Outland/boss_doomlord_kazzak.cpp:265]: RegisterSpellScript

**Review by Kitzunu — APPROVED** [review]

**Comment by Nefertumm** [comment]
Fix the startup errors and check the codestyle task, once the tasks pass, I'll make a review

**Comment by IntelligentQuantum** [comment]
:| AGAIN CONFLICT

**Comment by IntelligentQuantum** [comment]
@Nefertumm Done

**Comment by acidmanifesto** [comment]
Re-running workflow due to failure being unrelated to this PR

**Review by acidmanifesto — CHANGES_REQUESTED** [review]
certain modules will need to be checked as workflow points them out
image

**Comment by acidmanifesto** [comment]
merge conflict

**Review by acidmanifesto — APPROVED** [review]

**Comment by Kitzunu** [comment]
conflicts

**Comment by Yehonal** [comment]
Does it also close it? https://github.com/azerothcore/azerothcore-wotlk/issues/4209

**Pull request merged**